http/src/v0/root_files/add.rs

```diff
@@ -1,111 +1,307 @@
 use super::AddArgs;
 use crate::v0::support::StringError;
-use bytes::{Buf, Bytes};
+use bytes::{
+    buf::{BufExt, BufMutExt},
+    Buf, BufMut, Bytes, BytesMut,
+};
 use cid::Cid;
-use futures::stream::{Stream, TryStreamExt};
-use ipfs::{Ipfs, IpfsTypes};
+use futures::stream::{Stream, StreamExt, TryStreamExt};
+use ipfs::unixfs::ll::{
+    dir::builder::{
+        BufferingTreeBuilder, TreeBuildingFailed, TreeConstructionFailed, TreeNode, TreeOptions,
+    },
+    file::adder::FileAdder,
+};
+use ipfs::{Block, Ipfs, IpfsTypes};
 use mime::Mime;
-use mpart_async::server::MultipartStream;
+use mpart_async::server::{MultipartError, MultipartStream};
 use serde::Serialize;
 use std::borrow::Cow;
 use std::fmt;
 use warp::{Rejection, Reply};
 
 pub(super) async fn add_inner<T: IpfsTypes>(
     ipfs: Ipfs<T>,
-    _opts: AddArgs,
+    opts: AddArgs,
     content_type: Mime,
-    body: impl Stream<Item = Result<impl Buf, warp::Error>> + Unpin,
+    body: impl Stream<Item = Result<impl Buf, warp::Error>> + Send + Unpin + 'static,
 ) -> Result<impl Reply, Rejection> {
-    // FIXME: this should be without adder at least
-    use ipfs::unixfs::ll::file::adder::FileAdder;
-
     let boundary = content_type
         .get_param("boundary")
         .map(|v| v.to_string())
         .ok_or_else(|| StringError::from("missing 'boundary' on content-type"))?;
 
-    let mut stream =
-        MultipartStream::new(Bytes::from(boundary), body.map_ok(|mut buf| buf.to_bytes()));
+    let st = MultipartStream::new(Bytes::from(boundary), body.map_ok(|mut buf| buf.to_bytes()));
 
-    // this should be a while loop but clippy will warn if this is a while loop which will only get
-    // executed once.
-    if let Some(mut field) = stream
-        .try_next()
-        .await
-        .map_err(|e| StringError::from(format!("IO error: {}", e)))?
-    {
-        let field_name = field
-            .name()
-            .map_err(|e| StringError::from(format!("unparseable headers: {}", e)))?;
+    let st = add_stream(ipfs, st, opts);
+
+    // map the errors into json objects at least as we cannot return them as trailers (yet)
 
-        if field_name != "file" {
-            return Err(StringError::from(format!("unsupported field: {}", field_name)).into());
+    let st = st.map(|res| match res {
+        passthrough @ Ok(_) | passthrough @ Err(AddError::ResponseSerialization(_)) => {
+            // there is nothing we should do or could do for these; the assumption is that hyper
+            // will send the bytes and stop on serialization error and log it. the response
+            // *should* be closed on the error.
+            passthrough
         }
+        Err(something_else) => {
+            let msg = crate::v0::support::MessageResponseBuilder::default()
+                .with_message(format!("{}", something_else));
```

Comment:
```suggestion
                .with_message(something_else.to_string());
```

Reply:
have to admit I didn't hesitate to write that `format!("{}", ...)` a second :)